### DISCUSSION TITLE
HTTP Request 라이브러리 비교: Ky vs. Axios

### DISCUSSION THREAD

**Discussion by KingNono1030** [DevFDev/frontend#29] [학습 자료 공유]
# 📊 HTTP Request 라이브러리 비교: Ky vs. Axios (TanStack Query와 함께 사용하기)

### Axios
- **장점**:
  - **커뮤니티와 문서 지원**: 널리 사용되는 라이브러리로 참고 자료가 풍부함.
  - **광범위한 기능 제공**: 인터셉터, 요청 취소, 기본 설정(인스턴스) 생성, 응답 변환 등 다양한 기능을 제공하여 복잡한 요청을 쉽게 다룸.
  - **Node.js 지원**: 서버 사이드 코드에서도 사용하기 좋음.
- **단점**:
  - **무거운 패키지**: Ky에 비해 상대적으로 패키지 크기가 크며, 구형 Fetch API 기반이므로 최신 환경과의 호환성에서 Ky에 뒤짐.

### Ky
- **장점**:
  - **경량성과 최신성**: Fetch API를 기반으로 하여 작고 빠르며, 최신 브라우저 환경에서 최적화되어 있음.
  - **간단한 API**: 직관적이고 간단한 사용법으로 코드 작성이 쉬움.
  - **내장 기능 제공**: `retry`, `timeout`, `JSON 자동 파싱` 등의 기능을 기본 제공하여 코드가 간결함.
- **단점**:
  - **기능 제한**: 인터셉터와 같은 고급 기능이 없고, 브라우저 전용으로 설계되어 Node.js 환경에서 사용하기 어려움.

---

### 언제 Ky가 유리한가?
- Ky는 최신 Fetch API 기반의 가벼운 패키지로서 **경량화와 최신 환경 호환성**이 중요시되는 현대적인 프론트엔드 애플리케이션에 최적화되어 있습니다.
- 특히 React와 Vue 등에서 간단한 HTTP 요청이나 TanStack Query와 결합해 사용하기에 적합합니다.
- Ky는 기본 기능(재시도, 타임아웃 등)을 내장하고 있어 간단한 설정만으로도 충분한 HTTP 요청 구성을 할 수 있습니다.

### 언제 Axios가 더 적합한가?
- 복잡한 인터셉터 설정이나 다양한 환경(브라우저와 Node.js)을 모두 지원해야 하는 경우, Axios가 더 유리할 수 있습니다.
- 서버 사이드와 클라이언트 모두에서 동일한 HTTP 요청 구성을 사용하거나, 고급 기능이 필요할 때 적합합니다.

---

### TanStack Query와 Ky를 함께 사용하기 좋은 이유

1. **경량성**: Ky는 Axios보다 가볍고 단순하여 TanStack Query의 상태 관리 및 캐싱 기능과 잘 어우러집니다.
2. **단순하고 직관적인 사용법**: TanStack Query와 Ky 모두 간결한 API 구조를 갖추고 있어 코드 가독성이 높아집니다.
3. **기본 제공 기능과의 시너지**:
   - Ky의 기본 기능(`retry`, `timeout`, JSON 자동 파싱 등)이 TanStack Query의 재시도 기능과 함께 코드의 복잡성을 줄여줍니다.
   - TanStack Query가 요청을 캐싱하고 데이터 상태를 관리하는 동안 Ky는 가벼운 API로 데이터를 효율적으로 요청할 수 있습니다.

### 예시 코드 (TanStack Query + Ky)

```javascript
import ky from 'ky';
import { useQuery } from '@tanstack/react-query';

const fetchData = async () => {
  return await ky.get('https://api.example.com/data').json();
};

function MyComponent() {
  const { data, error, isLoading } = useQuery(['data'], fetchData);

  if (isLoading) return <p>Loading...</p>;
  if (error) return <p>Error: {error.message}</p>;

  return <div>{JSON.stringify(data)}</div>;
}
```

---

## 4. 최종 추천

1. 다양한 기능과 범용성이 중요한 경우: Axios
- 복잡한 요청 구성이나 서버 사이드 통합이 필요한 경우 Axios가 더 적합합니다.
2. 경량성과 최신 환경 최적화가 중요한 경우: Ky
- 최신 브라우저에서 간단하고 가벼운 HTTP 요청을 수행하려면 Ky가 좋은 선택입니다.
- 특히 TanStack Query와 Ky의 조합은 프론트엔드 전용 프로젝트에서 코드 단순화와 성능 최적화에 유리합니다.

---

## 참고 문서 요약

- [Why You Should Switch from axios to ky for Your Projects](https://medium.com/@muzammilsyed270300/why-you-should-use-ky-instead-of-axios-for-http-requests-in-your-frontend-2c7878be3b30)
- [Why Ky is the Best Alternative to Axios and Fetch](https://dev.to/usluer/why-ky-is-the-best-alternative-to-axios-and-fetch-for-modern-http-requests-27c3)

**Comment by KingNono1030** [comment]
@yongb2n 
axios 와 ky 용도의 차이, 장단점이 있어

**Comment by yellowjang** [comment]
ky가 서버 컴포넌트와의 호환성도 좋고, fetch의 단점들을 많이 보완해준다고 해서 사용해보면 좋을 것 같네요 ㅎㅎ
